include/pybind11/numpy.h

```diff
@@ -641,20 +641,23 @@ class dtype : public object {
             pybind11::str name;
             object format;
             pybind11::int_ offset;
+            field_descr(pybind11::str &&name, object &&format, pybind11::int_ &&offset)
+                : name{std::move(name)}, format{std::move(format)}, offset{std::move(offset)} {};
         };
+        auto field_dict = attr("fields").cast<dict>();
         std::vector<field_descr> field_descriptors;
+        field_descriptors.reserve(field_dict.size());
 
-        for (auto field : attr("fields").attr("items")()) {
-            auto spec = field.cast<tuple>();
-            auto name = spec[0].cast<pybind11::str>();
-            auto spec_fo = spec[1].cast<tuple>();
+        for (auto kv_item : field_dict) {
+            auto name = reinterpret_borrow<pybind11::str>(kv_item.first);
```

Comment:
IIUC this replaces `obj.cast<pybind11::str>()` — which does type checking — with `reinterpret_borrow<pybind11::str>` — which blindly assumes `kv_item.first` is a `str`. If that's true (?), and there is some upstream oversight/bug/API change, this could lead to a crash (relatively good) or a silent failure and time-consuming debugging.

Reply:
@rwgk Okay, I'll revert that part of the change. the dict_iterator returns an iterator of a std::pair<handles> which doesn't have a clean way of handling the steal / borrow distinction. I think you can call cast<> to a pyobject directly on it, but I don't know how it handles that code path to be honest.

Reply:
@rwgk I've reverted to the older dict api.